exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,229 @@
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    Eigen::MatrixXd Sk = Qf;
+    Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    for (int i = T - 2; i > 0; i--)
+    {
+        Eigen::MatrixXd x = prob_->get_X(i), u = prob_->get_U(i), Q = prob_->get_Q(i);
+        Eigen::MatrixXd Ak = dynamics_solver->fx(x, u), Bk = dynamics_solver->fu(x, u);
+
+        // eq. 13-16
+        Ak = Ak * dynamics_solver->get_dt() + Eigen::MatrixXd::Identity(Ak.rows(), Ak.cols());
+        Bk = Bk * dynamics_solver->get_dt();
+        // this inverse is common for all factors
+        Eigen::MatrixXd _inv = (Eigen::MatrixXd::Identity(R.rows(), R.cols()) * 1e-5 + R + Bk.transpose() * Sk * Bk).inverse();
+
+        Kv_gains_[i] = _inv * Bk.transpose();
+        K_gains_[i] = _inv * Bk.transpose() * Sk * Ak;
+        Ku_gains_[i] = _inv * R;
+        Sk = Ak.transpose() * Sk * (Ak - Bk * K_gains_[i]) + Q;
+
+        vk_gains_[i] = ((Ak - Bk * K_gains_[i]).transpose() * vk_gains_[i + 1]) -
+                     (K_gains_[i].transpose() * R * u) + (Q * x);
+
+        // fix for large values
+        Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+        vk_gains_[i] = vk_gains_[i].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+    }
+}
+
+double ILQRSolver::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    int T = prob_->get_T();
+    Eigen::VectorXd control_limits = prob_->GetControlLimits();
+
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+    for (int i = 0; i < T - 1; ++i)
+    {
+        Eigen::MatrixXd u = prob_->get_U(i);
+        // eq. 12
+        Eigen::MatrixXd delta_uk = -Ku_gains_[i] * u - Kv_gains_[i] * vk_gains_[i + 1] - K_gains_[i] * dynamics_solver->StateDelta(prob_->get_X(i), ref_trajectory.col(i));
+
+        u = u + alpha * delta_uk;
+        // clamp controls
+        if (control_limits.size() > 0)
+            u = u.cwiseMax(-control_limits).cwiseMin(control_limits);
+
+        prob_->Update(u, i);
+        cost += prob_->GetControlCost(i) + prob_->GetStateCost(i);
+    }
+
+    // add terminal cost
+    cost += prob_->GetStateCost(T - 1);
+    return cost;
+}
+
+void ILQRSolver::Solve(Eigen::MatrixXd& solution)
+{
+    int T = prob_->get_T();
+
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+
+    // all of the below are not pointers, since we want to copy over
+    //  solutions across iterations
+    Eigen::MatrixXd best_U, best_X, global_best_U;
+    solution.resize(Ku_gains_[0].rows(), T);
```

Comment:
I think the order should be the other way around. And please use `prob_->get_num_controls()` and `get_T()`.